### DISCUSSION TITLE
Extending QDMI_Job to avoid redundant calls

### DISCUSSION THREAD

**Discussion by kayaercument** [Munich-Quantum-Software-Stack/QDMI#60] [Ideas]
Firstly, please see @burgholzer 's [comment](https://github.com/Munich-Quantum-Software-Stack/QDMI/issues/39#issuecomment-2232919830).

I'm working on creating a minimal working backend for AQT's Ion Trap QPU. Even though I use Python binding to create AQT Job, it would be safe to assume that they need to iterate the bitcode to create AQT Job. Since there are function attributes such as `required_num_qubits` and `required_num_results`, they can be used to calculate `readout_size`.

**Comment by echavarria-mqv** [comment]
Hi @kayaercument, awesome job bringing this new QDMI backend to life! This will ensure a more tightened collaboration between AQT and the LRZ.

I need you to help me understand though, you mentioned `required_num_qubits` and `required_num_results`. These are attributes specific to QIR, correct? What exactly are you proposing? Are you suggesting modifying the QDMI specification, perhaps by altering the `QDMI_Job` struct, to account for these attributes? If so, kindly keep in mind that we must ensure QDMI remains as independent from QIR as possible. Other HPC centers adopting QDMI might prefer using a different intermediate representation, which would make QDMI effectively incompatible if we tailor it exclusively for QIR.

**Reply by kayaercument** [reply]
Hello @echavarria-lrz,

It does not have to be depending on the QIR. The same information can be extracted from QASM. 

I propose to alter `QDMI_Job` to include the following information: 

- Required number of qubits
- Required number of results/classical registers
- Number of shots. 

Including those information would be useful in the readout routines. For example, let's assume the following circuit is submitted to the backend.
```
define void @circuit-192() #0 {
entry:
  call void @__quantum__rt__initialize(i8* null)
  call void @__quantum__qis__h__body(%Qubit* null)
  call void @__quantum__qis__cnot__body(%Qubit* null, %Qubit* inttoptr (i64 1 to %Qubit*))
  call void @__quantum__qis__mz__body(%Qubit* null, %Result* null)
  call void @__quantum__qis__mz__body(%Qubit* inttoptr (i64 1 to %Qubit*), %Result* inttoptr (i64 1 to %Result*))
  call void @__quantum__rt__array_record_output(i64 2, i8* null)
  call void @__quantum__rt__result_record_output(%Result* inttoptr (i64 1 to %Result*), i8* null)
  call void @__quantum__rt__result_record_output(%Result* null, i8* null)
  ret void
}

attributes #0 = { "entry_point" "output_labeling_schema" "qir_profiles"="custom" "required_num_qubits"="2" "required_num_results"="2" }
```

And the response would be something like this:
```
{"job":{"job_type":"quantum_circuit","label":"xyxyx","job_id":"xyxyxyxy","resource_id":"","workspace_id":""},"response":{"status":"finished","result":{"0":[[1,1],[1,1],[0,0],[0,0],[0,0],[1,1],[1,1],[1,1],[1,1],[1,1],[1,1],[1,1],[1,1],[1,1],[1,1],[0,0],[1,1],[1,1],[0,0],[1,1],[1,1],[1,1],[1,1],[1,1],[0,0],[0,0],[0,0],[1,1],....]}}}
```

The `QDMI_control_readout_size` needs to query the job from API and parse the response to get the information. We can skip all the calls if we extract the information from the IR.
